File: TextEditorProject/TextEditorProject/TextEditorProject/TextEditor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TextEditorProject
{
    public class TextEditor
    {
        private string Content;
        private int CursorPosition;
        private List<string> Selections;

        public TextEditor()
        {
            Content = string.Empty;
            CursorPosition = 0;
            Selections = new List<string>();
        }

        public void SetContent(string content)
        {
            if (content == null)
            {
                Content = string.Empty;
            }
            else
            {
                this.Content = content;
            }
        }
        public void SetCursorPosition(int position)
        {
            if (position < 0 || position > Content.Length)
            {
                Console.WriteLine("Cursor is out of range");
            }
            else
            {
                CursorPosition = position;
            }
        }
        public void AddSelection(string selection)
        {
            Selections.Add(selection);
        }

        public void ClearSelections()
        {
            Selections.Clear();
        }
        public void DisplayState()
        {
            Console.WriteLine($"{Content}");
            Console.WriteLine($"{CursorPosition}");
            Console.WriteLine($"{Selections}");
        }

        public EditorMemento CreateMemento()
        {
            return new EditorMemento(Content, CursorPosition,Selections);
        }

        public void RestoreFromMemento(EditorMemento memento)
        {
            Content = memento.Content;
            CursorPosition = memento.CursorPosition;
            Selections = new List<string>(memento.Selections);
        }
    }
}

File: Lab-12/TextEditorProject/src/TextEditor.java
import java.util.ArrayList;
import java.util.List;

public class TextEditor {
    private String content;
    private int cursorPosition;
    private List<String> selections;

    public TextEditor() {
        this.content = "";
        this.cursorPosition = 0;
        this.selections = new ArrayList<>();
    }

    public void setContent(String content) {
        this.content = content;
    }

    public void setCursorPosition(int position) {

            this.cursorPosition = position;

    }

    public void addSelection(String selection) {
        selections.add(selection);
    }

    public void clearSelections() {
        selections.clear();
    }

    public void displayState() {
        System.out.println("Content: " + content);
        System.out.println("Cursor Position: " + cursorPosition);
        System.out.println("Selections: " + selections);
    }
}